azure-core/azure/core/pipeline/transport/requests_asyncio.py

```diff
@@ -0,0 +1,169 @@
+
+from .base import HttpRequest
+from .base_async import AsyncHttpTransport, AsyncHttpResponse
+from .requests_basic import RequestsTransport, RequestsTransportResponse
+from azure.core.exceptions import (
+    ServiceRequestError,
+    ConnectionError,
+    ConnectionReadError,
+    ReadTimeoutError,
+    raise_with_traceback
+)
+
+
+_LOGGER = logging.getLogger(__name__)
+
+
+def _get_running_loop():
+    try:
+        return asyncio.get_running_loop()
+    except AttributeError:  # 3.5 / 3.6
+        loop = asyncio._get_running_loop()  # pylint: disable=protected-access
+        if loop is None:
+            raise RuntimeError('No running event loop')
+        return loop
+
+
+class AsyncioRequestsTransport(RequestsTransport, AsyncHttpTransport):  # type: ignore
+
+    async def __aenter__(self):
+        return super(AsyncioRequestsTransport, self).__enter__()
+
+    async def __aexit__(self, *exc_details):  # pylint: disable=arguments-differ
+        return super(AsyncioRequestsTransport, self).__exit__()
+
+    async def send(self, request: HttpRequest, **kwargs: Any) -> AsyncHttpResponse:  # type: ignore
+        """Send the request using this HTTP sender.
+        """
+        loop = kwargs.get("loop", _get_running_loop())
+        response = None
+        error = None
+        if self.config.proxy_policy and 'proxies' not in kwargs:
+            kwargs['proxies'] = self.config.proxy_policy.proxies
+        try:
+            response = await loop.run_in_executor(
+                None,
+                functools.partial(
+                    self.session.request,
+                    request.method,
+                    request.url,
+                    headers=request.headers,
+                    data=request.data,
+                    files=request.files,
+                    verify=kwargs.get('connection_verify', self.config.connection.verify),
+                    timeout=kwargs.get('connection_timeout', self.config.connection.timeout),
+                    cert=kwargs.get('connection_cert', self.config.connection.cert),
+                    allow_redirects=False,
+                    **kwargs))
+
+        except urllib3.exceptions.NewConnectionError as err:
+            error = ConnectionError(err, error=err)
+        except requests.exceptions.ReadTimeout as err:
+            error = ReadTimeoutError(err, error=err)
+        except requests.exceptions.ConnectionError as err:
+            if err.args and isinstance(err.args[0], urllib3.exceptions.ProtocolError):
+                error = ConnectionReadError(err, error=err)
+            else:
+                error = ConnectionError(err, error=err)
+        except requests.RequestException as err:
+            error = ServiceRequestError(err, error=err)
+        finally:
+            if not self.config.connection.keep_alive and (not response or not kwargs['stream']):
+                self.session.close()
+        if error:
+            raise error
```

Comment:
You are Python 3 here, use a "raise from" instead

Reply:
hmm yes I that's true - however I'm curious about your thoughts on stack traces....

If we use raise from, we get the full stack trace - which can be useful, but not always, and can end up being extremely long.
The way I've written it here is to effectively cut the stack trace so we don't see all the methods in requests, urllib3 etc. I've then added that detailed trace back as a property on the error:
https://github.com/Azure/azure-sdk-for-python/blob/azure-core/azure-core/azure/core/exceptions.py#L63

Along with a "raise_with_traceback" function on the error for those who want to raise with the full trace back:
https://github.com/Azure/azure-sdk-for-python/blob/azure-core/azure-core/azure/core/exceptions.py#L69

Again - I don't know what the right thing to do here is. But we were getting some complaints around the length of the trace backs - so this is my first pass at a workaround :)